webauthn4j-core/src/main/java/com/webauthn4j/validator/attestation/statement/tpm/TPMAttestationStatementValidator.java

```diff
@@ -293,57 +270,56 @@ private void validateSubjectAlternativeName(X509Certificate certificate) throws
-        return ByteBuffer.allocate(authenticatorData.length + clientDataHash.length).put(authenticatorData).put(clientDataHash).array();
-    }
+    TPMDeviceProperty parseTPMDeviceProperty(LdapName directoryName) throws IOException, NamingException {
+        Map<String, byte[]> map = directoryName.getRdns().stream()
+                .map(item -> convertAttributesToMap(item.toAttributes()).entrySet())
+                .flatMap(Collection::stream)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
 
+        byte[] manufacturerAttr = map.get("2.23.133.2.1");
+        byte[] partNumberAttr = map.get("2.23.133.2.2");
+        byte[] firmwareVersionAttr = map.get("2.23.133.2.3");
 
-    private List<KeyValue> toKeyValue(Attributes attributes)  {
-        NamingEnumeration<String> attributesIDs = attributes.getIDs();
-        //Map<String, Object> result = new HashMap<>();
-        List<KeyValue> result = new ArrayList<>();
+        String manufacturer = decodeAttr(manufacturerAttr);
+        String partNumber = decodeAttr(partNumberAttr);
+        String firmwareVersion = decodeAttr(firmwareVersionAttr);
 
-        try {
-            while (attributesIDs.hasMore()) {
-                String aid = attributesIDs.next();
-                Object value = attributes.get(aid).get(); //casting to byte is a bad idea?
+        return new TPMDeviceProperty(manufacturer, partNumber, firmwareVersion);
+    }
 
-                //result.put(aid, value);
-                result.add(new KeyValue(aid, value));
-            }
-        } catch (Exception e) {
-            //
+    private String decodeAttr(byte[] attr) throws IOException {
+        String value = null;
+        if(attr != null){
+            Asn1Utf8String partNumberUtf8String = new Asn1Utf8String();
+            partNumberUtf8String.decode(attr);
+            value = partNumberUtf8String.getValue();
         }
-
-        return result;
+        return value;
     }
 
-    class KeyValue {
-        private String key;
-        private Object value;
-
-        KeyValue(String key, Object value) {
-            this.key = key;
-            this.value = value;
-        }
+    private Map<String, byte[]> convertAttributesToMap(Attributes attributes) {
+        try{
+            NamingEnumeration<String> attributesIDs = attributes.getIDs();
+            Map<String, byte[]> result = new HashMap<>();
+            while (attributesIDs.hasMore()) {
+                String aid = attributesIDs.next();
+                byte[] value = (byte[])attributes.get(aid).get();
 
-        public String getKey() {
-            return key;
+                result.put(aid, value);
+            }
+            return result;
         }
-
-        public void setKey(String key) {
-            this.key = key;
+        catch (NamingException e){
+            throw new BadAttestationStatementException("Failed to parse attestation certificate", e);
```

Comment:
Change the exception message to "Failed to parse the TPM attestation attributes"?

Reply:
Thank you for suggestion. will correct.